packages/health/ios/Sources/SwiftHealthPlugin+Data.swift

```diff
@@ -136,125 +136,148 @@ extension SwiftHealthPlugin {
+      withStart: input.dateFrom,
+      end: input.dateTo,
       options: .strictStartDate
     )
     let sort = NSSortDescriptor(
       key: HKSampleSortIdentifierEndDate,
       ascending: false
     )
-
     let query =  HKSampleQuery(
-      sampleType: sampleType,
+      sampleType: input.sampleType,
       predicate: predicate,
-      limit: limit,
-      sortDescriptors: [sort]
-    ) { _, samples, error in
-
-      if let error {
-        logger.error("\(#function) query failed: \(error.localizedDescription)")
-        DispatchQueue.main.async {
-          result(nil)
+      limit: input.limit,
+      sortDescriptors: [sort],
+      resultsHandler: { _, samples, error in
+        if let error {
+          logger.error("\(#function) query failed: \(error.localizedDescription)")
+          onCompletion(.failure(PluginError(message: error.localizedDescription)))
+          return
         }
-        return
-      }
-
-      guard let samples else {
-        DispatchQueue.main.async {
-          result(nil)
+        guard let samples else {
+          onCompletion(.success([]))
+          return
         }
-        return
+        Self.parseGetDataQueryResults(
+          type: input.type,
+          unit: input.unit,
+          samples: samples,
+          onCompletion: onCompletion
+        )
       }
+    )
+    healthStore.execute(query)
+  }
 
-      if let samples = samples as? [HKQuantitySample] {
-        let dictionaries = samples.map { sample -> NSDictionary in
-          return [
-            "uuid": "\(sample.uuid)",
-            "value": sample.quantity.doubleValue(for: unit),
-            "date_from": Int(sample.startDate.timeIntervalSince1970 * 1000),
-            "date_to": Int(sample.endDate.timeIntervalSince1970 * 1000),
-            "source_id": sample.sourceRevision.source.bundleIdentifier,
-            "source_name": sample.sourceRevision.source.name,
-            "metadata": sample.metadata ?? [:],
-          ]
-        }
-        DispatchQueue.main.async { [dictionaries] in
-          result(dictionaries)
-        }
-        return
+  private static func parseGetDataQueryResults(
+    type: HealthDataTypes,
+    unit: HKUnit,
+    samples: [HKSample],
+    onCompletion: @escaping (Result<[[String: Any]], PluginError>) -> Void
+  ) {
+    if let samples = samples as? [HKQuantitySample] {
+      let dictionaries = samples.map { sample -> [String: Any] in
+        let metadata = sample.metadata?.reduce(into: [String: Any](), { dict, element in
+          if let value = element.value as? NSString {
+            dict[element.key] = String(value)
+          } else if let value = element.value as? NSNumber {
+            dict[element.key] = value.doubleValue
+          } else if let value = element.value as? NSDate {
+            dict[element.key] = value.timeIntervalSince1970 * 1000
+          } else {
+            logger.error("\(#function) Unexpected metadata type for \(type.rawValue) \(element.key)")
+          }
```

Comment:
prospective fix idea 1

Reply:
This is a good example where having remote non-fatals would be helpful (as in, would be nice to pull this into our swift and kotlin codebase directly)